### ISSUE TITLE
Separating Redis serialization from allowed types for column

### ISSUE THREAD

**Issue by peter-conalgo** [josiahcarlson/rom#109]
I want to create a column that does not allow itself to be set carelessly to the wrong type. The problem is that `Column.__set__` will call `self._from_redis(v)` for any type that's not otherwise allowed. This permits me to set a `rom.DateTime` column to 23 or '23' which I consider a type error.

I tried adding code in front of `__set__` on my column to do additional checking, however, at that level there is no way to distinguish database loading from user actions. It looks like `Model._init()` when loading, uses `setattr` so there is no way to pass in additional flags.

Perhaps there should a method on `Column` just for loading raw data from Redis.

**Comment by josiahcarlson** [comment]
This one is weird, but was actually doable on your side. The 99% solution is a custom column that overrides `_from_redis` to only accept `datetime` and `str` with specific formats. The 100% solution is a custom column that overrides `__set__` to handle checking for whether the instance is created yet or not, for determining whether to pass through `str` values for conversion.

That custom checking at set time could be made into an API... but it feels to me that Python comes with that already - subclassing. And rom supports subclassed/custom attribute types easily.

Really, the only reason that the DateTime object doesn't do this better is because Redis doesn't have a native datetime object, so to hack it, all datetimes are transformed into float seconds after the unix epoch (turned into a str on its way to Redis). Same with `datetime.date` and `datetime.time` objects (all time objects are assumed to occur on Jan 1, 1970).

For the fixing of both the `_from_redis` method as well as the `__set__` method, I added a `SaferDateTime` column type that has a new set of tests:

```
class RomTestCheckedColumns(Model):
    c = SaferDateTime()

# should be okay
RomTestCheckedColumns(c=datetime(1970, 1, 2)).save()

# also should work, because we need to round trip Redis
a = RomTestCheckedColumns(c='86400')
a.save()
# Fail on setting attributes
self.assertRaises(InvalidColumnValue, lambda: setattr(a, 'c', 1))
self.assertRaises(InvalidColumnValue, lambda: setattr(a, 'c', 'blah'))
self.assertRaises(InvalidColumnValue, lambda: setattr(a, 'c', '1'))
self.assertRaises(InvalidColumnValue, lambda: setattr(a, 'c', ()))
self.assertRaises(InvalidColumnValue, lambda: setattr(a, 'c', []))
self.assertRaises(InvalidColumnValue, lambda: setattr(a, 'c', {}))

# Should fail!
self.assertRaises(InvalidColumnValue, lambda: RomTestCheckedColumns(c=86400))
```

For your needs (now that you know that the underlying representation is float seconds since the epoch), you can change your underlying column representation to be more resistant to weird data (see `_to_redis` and `_from_redis`), while also doing the 'is it instantiated' check in `__set__`.

**Comment by josiahcarlson** [comment]
To see an example of what I'm doing with the check on instantiation: https://github.com/josiahcarlson/rom/blob/master/rom/columns.py#L499

**Comment by josiahcarlson** [comment]
Actually, this one was more complicated than I thought, and there was a mistype. I've just released 0.39.3 with the fixed interface, type checking, and fixed tests https://github.com/josiahcarlson/rom/blob/master/rom/columns.py#L469

**Comment by peter-conalgo** [comment]
```python
class Testing(rom.Model):
    foo = rom.SaferDateTime()
```

`Testing(foo='123')` fails, and so does `Testing(foo=123)` which is good. However `Testing()` and `Testing(foo=None)` fail as well (with `InvalidColumnValue`).

My example above about dates was meant as more of an instance of the problem, not a column type that I want. [Aside: IMHO a superior datetime column would store ISO8601 values and use a specialized numeric keygen for ordering.]

But your answer does provide a clue to what I need: I can check for database loading case in `__set__` by considering `_init` attribute of the object being set.
